File: algorithms/php/42.TrappingRainWater.php
<?php

class Solution
{

    /**
     * @param Integer[] $height
     * @return Integer
     */
    function trap($height)
    {
        $count = count($height);
        $max_left = $max_right = array_fill(0, $count, 0);

        for ($i = 1; $i < $count; $i++) {
            $max_left[$i] = max($max_left[$i - 1], $height[$i - 1]);
        }

        for ($i = $count - 2; $i >= 0; $i--) {
            $max_right[$i] = max($max_right[$i + 1], $height[$i + 1]);
        }

        $sum = 0;
        for ($i = 0; $i < $count; $i++) {
            if (min($max_left[$i], $max_right[$i]) > $height[$i]) {
                $sum += min($max_left[$i], $max_right[$i]) - $height[$i];
            }
        }

        return $sum;
    }
}

File: algorithms/cpp/42.TrappingRainWater.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int count = height.size();
        vector<int> maxLeft(count, 0);
        vector<int> maxRight(count, 0);

        for (int i = 1; i < count; i++) {
            maxLeft[i] = max(maxLeft[i - 1], height[i - 1]);
        }

        for (int i = count - 2; i >= 0; i--) {
            maxRight[i] = max(maxRight[i + 1], height[i + 1]);
        }

        int sum = 0;
        for (int i = 0; i < count; i++) {
            if (min(maxLeft[i], maxRight[i]) > height[i]) {
                sum += min(maxLeft[i], maxRight[i]) - height[i];
            }
        }

        return sum;
    }
};
